### PR TITLE
clientv3: fix Close after failed Put

### PR DESCRIPTION
**Pull request by heyitsanthony**
Was crashing on a nil connection. Reworked the shutdown path a little so
there's only one connection close site.

### CODE REVIEW

**Comment by gyuho** [comment]
LGTM. Defer to @xiang90

File: clientv3/integration/kv_test.go [hunk 1]
```diff
@@ -547,3 +547,13 @@ func TestKVGetCancel(t *testing.T) {
...
 }
+
+// TestKVPutStoppedServerAndClose ensures closing after a failed Put works.
+func TestKVPutStoppedServerAndClose(t *testing.T) {
+	t.Parallel()
+	defer testutil.AfterTest(t)
+	clus := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+	clus.Members[0].Stop(t)
+	clus.Client(0).Put(context.TODO(), "abc", "123")
```
Comment (xiang90) [inline clientv3/integration/kv_test.go]: Add the comment to tell that this test depends on afterTest to check leak? Or it is unclear to me why we just do a put without checking anything.

File: clientv3/client.go [hunk 2]
```diff
@@ -213,16 +217,16 @@ func (c *Client) retryConnection(err error) (newConn *grpc.ClientConn, dialErr e
...
-	}
 	if c.conn != nil {
 		c.conn.Close()
 		if st, _ := c.conn.State(); st != grpc.Shutdown {
 			// wait so grpc doesn't leak sleeping goroutines
-			c.conn.WaitForStateChange(c.ctx, st)
+			c.conn.WaitForStateChange(context.Background(), st)
 		}
 	}
+	if c.cancel == nil {
```
Comment (xiang90) [inline clientv3/client.go]: I feel we need to add a comment about this check. cancel does not tell the client is closed explicitly.

File: clientv3/client.go [hunk 3]
```diff
@@ -262,8 +266,9 @@ func (c *Client) connMonitor() {
 		select {
 		case err = <-c.reconnc:
 		case <-c.ctx.Done():
+			_, err = c.retryConnection(c.ctx.Err())
```
Comment (xiang90) [inline clientv3/client.go]: this is just for appending the error?

Comment (heyitsanthony) [inline clientv3/client.go]: it also closes the connection, if any

File: clientv3/integration/kv_test.go (hunk 1 above)
Comment (heyitsanthony) [inline clientv3/integration/kv_test.go]: ok added comment

File: clientv3/client.go (hunk 2 above)
Comment (heyitsanthony) [inline clientv3/client.go]: ok commented

**Comment by xiang90** [comment]
lgtm

**Pull request merged**